ksql-metastore/src/main/java/io/confluent/ksql/metastore/model/StructuredDataSource.java

```diff
@@ -13,42 +13,50 @@
  * specific language governing permissions and limitations under the License.
  */
 
-package io.confluent.ksql.metastore;
+package io.confluent.ksql.metastore.model;
 
-import io.confluent.ksql.query.QueryId;
+import static java.util.Objects.requireNonNull;
+
+import com.google.errorprone.annotations.Immutable;
+import io.confluent.ksql.metastore.SerdeFactory;
+import io.confluent.ksql.schema.ksql.KsqlSchema;
 import io.confluent.ksql.serde.DataSource;
 import io.confluent.ksql.serde.KsqlTopicSerDe;
 import io.confluent.ksql.util.timestamp.TimestampExtractionPolicy;
+import java.util.Optional;
 import org.apache.kafka.connect.data.Field;
 import org.apache.kafka.connect.data.Schema;
 
-public abstract class StructuredDataSource implements DataSource {
-
-  protected final String dataSourceName;
-  protected final DataSourceType dataSourceType;
-  protected final Schema schema;
-  protected final Field keyField;
-  protected final TimestampExtractionPolicy timestampExtractionPolicy;
+@Immutable
+public abstract class StructuredDataSource<K> implements DataSource {
```

Comment:
I can't seem to find any usage of this class that doesn't specify `<?>`. Whats the point of having the type parameter at all?

Reply:
Oh I'm well aware!  This is on my radar. However, I'm not trying to fix that now, just refactor the key from the two derived types into the base. Hence making this templated like the two derived types.